### ISSUE TITLE
[Bug] Potential cache consistency issues in multi-threaded env

### ISSUE THREAD

**Issue by bgavrilMS** [AzureAD/microsoft-authentication-library-for-dotnet#1685]
**Which Version of MSAL are you using ?**
4.9

**Platform**
.net and .net core

**What authentication flow has the issue?**
All, but confidential flows affected the most

**Repro**

In high load enviroments, where the token cache is shared by many threads, cache inconsistency can occur because the access to the underlying token cache is not fully syncronized. In particular, the in-memory cache uses normal dictionaries, which are not thread safe if read and written to at the same time.

This could become a problem if the same token cache is shared between multiple threads (singleton CCA in web apps) or processes (dev tools cache).

Note: we've not yet had users complaining about this.

**Comment by bgavrilMS** [comment]
Example problematic call: 

```csharp
            await RefreshCacheForReadOperationsAsync(CacheEvent.TokenTypes.Account).ConfigureAwait(false);
            return await TokenCacheInternal.GetAccountsAsync(authority, _requestParams.RequestContext).ConfigureAwait(false);

```

**Comment by GGorelko** [comment]
The issues I'm currently facing with similarly are related to the bug above (package version 4.9). The web app is spinning up multiple threads to send requests acquiring access tokens for Confidential client and I can see from time to time exceptions like below (any hints how to work around this)?:

> 
> System.AggregateException: One or more errors occurred. ---> System.ArgumentNullException: Value cannot be null.
> Parameter name: source
>    at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
>    at Microsoft.Identity.Client.TelemetryCore.TelemetryManager.Flush(String correlationId)
>    at Microsoft.Identity.Client.Internal.Requests.RequestBase.<RunAsync>d__14.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.<ExecuteAsync>d__3.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at MS.Dynamics.Infrastructure.Gates.Gtc.CloudBuild.CloudBuildAadAuth.<GetAccessToken>d__9.MoveNext() in F:\One\Dynamics\Infrastructure\Gates\src\Services\Gtc\GtcTaskProvider.CloudBuild\CloudBuildAadAuth.cs:line 77
>    --- End of inner exception stack trace ---

           `ConfidentialClientApplicationBuilder appBuilder = 
            ConfidentialClientApplicationBuilder.Create(_applicationId);
            appBuilder.WithTenantId(_tenantId);
            appBuilder.WithClientSecret(_clientSecret);             

            IConfidentialClientApplication application = appBuilder.Build();
            AcquireTokenForClientParameterBuilder tokenAcquirer = application.AcquireTokenForClient(new List<string> {_scope});
            AuthenticationResult authResult = await tokenAcquirer.ExecuteAsync();
            string accessToken = authResult.AccessToken; `

**Comment by bgavrilMS** [comment]
@GGorelko - could you log a separate issue for this please? We treat null reference exceptions as high priority bugs.

**Comment by neha-bhargava** [comment]
Fixed in [MSAL 4.12.0 release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/4.12.0)
